openpype/hosts/nuke/plugins/publish/collect_writes.py

```diff
@@ -33,112 +28,93 @@ def process(self, instance):
 
-            self._add_farm_instance_data(instance)
+        if render_target == "farm":
+            self.add_farm_instance_data(instance)
 
-        elif render_target == "farm":
-            self._add_farm_instance_data(instance)
-
-        # set additional instance data
-        self._set_additional_instance_data(instance, render_target, colorspace)
-
-    def _set_existing_files_data(self, instance, colorspace):
-        """Set existing files data to instance data.
-
-        Args:
-            instance (pyblish.api.Instance): pyblish instance
-            colorspace (str): colorspace
-
-        Returns:
-            list: collected frames
-        """
-        collected_frames = self._get_collected_frames(instance)
-
-        representation = self._get_existing_frames_representation(
-            instance, collected_frames
-        )
+        elif render_target == "frames":
+            file_paths = self.prepare_collection_of_file_paths(
+                write_file_path, frame_start, frame_end, only_existing=True
+            )
 
-        # inject colorspace data
-        self.set_representation_colorspace(
-            representation, instance.context,
-            colorspace=colorspace
-        )
+            representation = self.prepare_representation(
+                instance, file_paths, frame_start, frame_end
+            )
 
-        instance.data["representations"].append(representation)
+            # QUESTION: should we set colorspace at this moment or downstream?
+            self.set_representation_colorspace(
+                representation, instance.context,
+                colorspace=colorspace
+            )
+            instance.data["representations"].append(representation)
 
-        return collected_frames
+        elif render_target == "frames_farm":
+            file_paths = self.prepare_collection_of_file_paths(
+                write_file_path, frame_start, frame_end, only_existing=True
+            )
 
-    def _set_expected_files(self, instance, collected_frames):
-        """Set expected files to instance data.
+            representation = self.prepare_representation(
+                instance, file_paths, frame_start, frame_end
+            )
+            self.make_farm_publishing_representation(representation)
 
-        Args:
-            instance (pyblish.api.Instance): pyblish instance
-            collected_frames (list): collected frames
-        """
-        write_node = self._write_node_helper(instance)
+            # QUESTION: should we set colorspace at this moment or downstream?
+            self.set_representation_colorspace(
+                representation, instance.context,
+                colorspace=colorspace
+            )
+            instance.data["representations"].append(representation)
 
-        write_file_path = nuke.filename(write_node)
-        output_dir = os.path.dirname(write_file_path)
+            self.add_farm_instance_data(instance)
```

Comment:
Agree on this it could be just
```python
if render_target in {"frame", "frames_farm"}:
    file_paths = self.prepare_collection_of_file_paths(
        write_file_path, frame_start, frame_end, only_existing=True
    )

    representation = self.prepare_representation(
        instance, file_paths, frame_start, frame_end
    )

    # QUESTION: should we set colorspace at this moment or downstream?
    self.set_representation_colorspace(
        representation, instance.context,
        colorspace=colorspace
    )
    instance.data["representations"].append(representation)

if render_target in {"farm", "frames_farm"}:
    self.add_farm_instance_data(instance)
```

Reply:
I'd personally like it actually if we just have a basic mixin we can use everywhere - similar to the extractor for `staging_dir` method that would expose e.g. the `add_representation` method and have all extractors use that instead. That way, in some time, if the data storage on the instances need to change we can just change whatever the method does.

Plus, having a dedicated function can also make sure the actual representation gets validated at that point in time, etc.

Anyway, having dedicated methods for that has come up quite a bit as a request actually. I believe there's a forum discussion.